doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,561 @@
+- **Output**: SDFG with transformed copies
+- **Current Location**: `cuda.py` preprocessing, various target preprocessors
+- **Applies To**: GPU strided copies, FPGA transfers
+
+#### 8. **StreamAssignmentPass** (GPU-specific)
+- **Purpose**: Assign CUDA streams for concurrent execution
+- **Input**: GPU-targeted SDFG
+- **Output**: SDFG with stream assignments, pipeline_results["stream_info"]
+- **Current Location**: Embedded in CUDA code generation
+
+#### 9. **TaskletLanguageLoweringPass**
+- **Purpose**: Convert Python/generic tasklets to target language (C++/CUDA/etc.)
+- **Input**: SDFG with tasklets
+- **Output**: SDFG with lowered tasklets
+- **Current Location**: Distributed across target generators
+
+### Phase 3: Code Generation Passes
+
+#### 10. **StateStructCreationPass**
+- **Purpose**: Generate state struct definitions for persistent data
+- **Input**: SDFG with allocation info
+- **Output**: pipeline_results["state_struct"] = {struct_def, struct_init}
+- **Current Location**: `DaCeCodeGenerator.generate_code()`
+
+#### 11. **AllocationCodePass**
+- **Purpose**: Generate allocation/deallocation code for each scope
+- **Input**: SDFG with allocation analysis
+- **Output**: pipeline_results["allocation_code"] = {alloc_code, dealloc_code}
+- **Current Location**: `allocate_arrays_in_scope()`, `deallocate_arrays_in_scope()`
+
+#### 12. **MemletLoweringPass**
+- **Purpose**: Lower high-level memlets to explicit copy operations
+- **Input**: SDFG with target analysis
+- **Output**: SDFG with explicit copies as tasklets
+- **Current Location**: Embedded in target-specific copy generation
+
+#### 13. **FrameCodeGenerationPass**
+- **Purpose**: Generate the main frame code with function signatures and dispatch
+- **Input**: All previous pass results
+- **Output**: pipeline_results["frame_code"] = {global_code, local_code}
+- **Current Location**: `DaCeCodeGenerator.generate_code()`
+
+#### 14. **TargetCodeGenerationPass**
+- **Purpose**: Generate target-specific code files
+- **Input**: SDFG with frame code
+- **Output**: List of CodeObject instances
+- **Current Location**: Target-specific `get_generated_codeobjects()`
+
+#### 15. **HeaderGenerationPass**
+- **Purpose**: Generate C/C++ header files for SDFG interface
+- **Input**: SDFG with frame code
+- **Output**: pipeline_results["headers"] = {call_header, sample_main}
+- **Current Location**: `generate_headers()`, `generate_dummy()`
+
+### Phase 4: File Generation Passes
+
+#### 16. **SDFGSplittingPass**
+- **Purpose**: Split complex SDFGs into multiple files if needed
+- **Input**: Single SDFG with all code
+- **Output**: List of SDFGs (one per target file)
+- **Current Location**: Implicit in current system
+
+#### 17. **CodeObjectCreationPass**
+- **Purpose**: Create final CodeObject instances for each generated file
+- **Input**: Split SDFGs and code
+- **Output**: List[CodeObject] ready for compilation
+- **Current Location**: End of `generate_code()`
```

Comment:
Should it be a pass or should it simply be the result of the new "code generation" pass (see above comment on splitting pass)?